cmd/deploy.go

```diff
@@ -49,13 +53,12 @@ func newDeployClient(cfg deployConfig) (*fn.Client, error) {
 
 	listener.Verbose = cfg.Verbose
 	builder.Verbose = cfg.Verbose
-	pusher.Verbose = cfg.Verbose
 	deployer.Verbose = cfg.Verbose
 
 	return fn.New(
 		fn.WithProgressListener(listener),
 		fn.WithBuilder(builder),
-		fn.WithPusher(pusher),
+		pusherOption,
```

Comment:
nit: wouldn't it be easier just to extract the `pusher` variable?

Reply:
I am not sure what exactly do you propose. But we need to cover 2 cases: 1. with push initialized and one basically with `nil` pusher.  The very same approach is being used for `build`, so I am keeping it consistent.
But if you have a better solution, I'd happy to use it. But I am not sure what do you propose 🤷‍♂️

Reply:
```diff
diff --git a/cmd/build.go b/cmd/build.go
index 16c897a5..b2f2f18a 100644
--- a/cmd/build.go
+++ b/cmd/build.go
@@ -29,13 +29,16 @@ func newBuildClient(cfg buildConfig) (*fn.Client, error) {
        builder := buildpacks.NewBuilder()
        listener := progress.New()
 
-       pusherOption := fn.WithPusher(nil)
+       var (
+               pusher *docker.Pusher
+               err    error
+       )
        if cfg.Push {
                credentialsProvider := creds.NewCredentialsProvider(
                        creds.WithPromptForCredentials(newPromptForCredentials()),
                        creds.WithPromptForCredentialStore(newPromptForCredentialStore()),
                        creds.WithTransport(cfg.Transport))
-               pusher, err := docker.NewPusher(
+               pusher, err = docker.NewPusher(
                        docker.WithCredentialsProvider(credentialsProvider),
                        docker.WithProgressListener(listener),
                        docker.WithTransport(cfg.Transport))
@@ -44,14 +47,13 @@ func newBuildClient(cfg buildConfig) (*fn.Client, error) {
                        return nil, err
                }
                pusher.Verbose = cfg.Verbose
-               pusherOption = fn.WithPusher(pusher)
        }
        builder.Verbose = cfg.Verbose
        listener.Verbose = cfg.Verbose
 
        return fn.New(
                fn.WithBuilder(builder),
-               pusherOption,
+               fn.WithPusher(pusher),
                fn.WithProgressListener(listener),
                fn.WithRegistry(cfg.Registry), // for image name when --image not provided
                fn.WithVerbose(cfg.Verbose)), nil
```

Reply:
why nil and not noop impl?

Reply:
gotcha

Reply:
or:
```diff
diff --git a/cmd/build.go b/cmd/build.go
index 16c897a5..4eae5aaa 100644
--- a/cmd/build.go
+++ b/cmd/build.go
@@ -29,7 +29,13 @@ func newBuildClient(cfg buildConfig) (*fn.Client, error) {
        builder := buildpacks.NewBuilder()
        listener := progress.New()
 
-       pusherOption := fn.WithPusher(nil)
+       opts := []fn.Option{
+               fn.WithBuilder(builder),
+               fn.WithProgressListener(listener),
+               fn.WithRegistry(cfg.Registry), // for image name when --image not provided
+               fn.WithVerbose(cfg.Verbose),
+       }
+       
        if cfg.Push {
                credentialsProvider := creds.NewCredentialsProvider(
                        creds.WithPromptForCredentials(newPromptForCredentials()),
@@ -44,17 +50,12 @@ func newBuildClient(cfg buildConfig) (*fn.Client, error) {
                        return nil, err
                }
                pusher.Verbose = cfg.Verbose
-               pusherOption = fn.WithPusher(pusher)
+               opts = append(opts, fn.WithPusher(pusher))
        }
        builder.Verbose = cfg.Verbose
        listener.Verbose = cfg.Verbose
 
-       return fn.New(
-               fn.WithBuilder(builder),
-               pusherOption,
-               fn.WithProgressListener(listener),
-               fn.WithRegistry(cfg.Registry), // for image name when --image not provided
-               fn.WithVerbose(cfg.Verbose)), nil
+       return fn.New(opts...), nil
 }
 
 type buildClientFn func(buildConfig) (*fn.Client, error)
 ```
 this ^^^ should use noop impl if it's desired

Reply:
also if the pusher is set to `nil` then push on client panics, shouldn't it rather return error?

Reply:
Fixed, that's much better. Thanks!

Reply:
>also if the pusher is set to nil then push on client panics, 

but that's only problem for test since in production code the `Push()` it's guarded.
well actually it's not problem for the test either since it's catching panic (which I find weird).

Reply:
Yeah, I have had the same concern, but I think that it is okay this way. The panic is being tested here: https://github.com/knative-sandbox/kn-plugin-func/pull/739/files#diff-0b3e07894eabba354682863bda0d8a2e0b5e8166e81e20d13923695b720874c4R145

Reply:
Well it tests that panics doesn't happen when it shouldn't, but it doesn't test that it does panic when it should.

Reply:
IMHO we should rather return error than crash.

Reply:
but that's Independent of this PR.

Reply:
+1